impress-org/givewp-next-gen
# Feature: add stateless donation confirmation receipt routing, template, and implementation

## Description



This introduces stateless receipts.  This means we will not be relying on the session to store information about our donation to display a confirmation receipt. 

Here are the most common scenarios for receipts:

1. Onsite gateways will redirect to the receipt within their iframe.
2. Offsite gateways will redirect back to the origin page of the donation form - which will load the receipt where the donation form was.

**Donation Confirmation Receipt View Route**

This url will mainly be used _under-the-hood_ to display the confirmation receipt in various scenarios.  Typically this means this will be the url of an iframe.  You can also use this route to view the receipt as a standalone page with nothing else on it.

```
$receiptId = $donation->purchaseKey

/?givewp-route=donation-confirmation-receipt-view&receipt-id={receiptId}
```


**Donation Confirmation Receipt Url**

This is for displaying the receipt in place of an embedded donation form.

```
{originUrl}/?givewp-event=donation-completed&givewp-listener=show-donation-confirmation-receipt&givewp-receipt-id={receiptId}&givewp-embed-id={embedId}
```


**Donation Flow & Controller Updates**

In order to preserve our framework's handler for `PaymentComplete` we are now filtering the default `give_get_success_page_uri` with our receipt view url.  This is primary used for onsite gateways.

Offsite gateways are more complicated but will now have an easy way to grab the redirectReturnUrl straight from `$gatewayData['redirectReturnUrl']`.  Once we start implementing _existing_ gateways into Next Gen we will filter this url on the legacy adapter so the gateways will always receive a value - but it will be determined by the request.


## Affects



This touches the next gen donation form flow including the current gateways and a new offsite test gateway.

## Visuals




<img width="434" alt="Screen Shot 2023-01-13 at 10 06 20 AM" src="https://user-images.githubusercontent.com/10138447/212352223-b6507b1d-13c6-4d4d-81ad-86b99e0f9e58.png">


## Testing Instructions



- Rebuild assets
- Add a new donation form block
- Enable Next Gen Gateways (_Next Gen Test Gateway, Next Gen Test Gateway Offsite, and Next Gen Stripe_)
- Donate using all gateways
- The form should be directed to new receipt after completion
- Now you can try adding custom fields to form to see them on receipt
- You can also add additional form blocks to the page - the receipt will only display in place of the one completed
- You can also visit the standalone receipt view url for any donation using the purchase ke which can be found in the "Donation Meta" box (labeled "Key") in the single donation admin. `/?givewp-route=donation-confirmation-receipt-view&receipt-id={key}`

## Pre-review Checklist
